crates/runtime/src/memory.rs

```diff
@@ -254,35 +260,136 @@ impl RuntimeLinearMemory for MmapMemory {
-        base: &'static mut [u8],
+/// A "static" memory where the lifetime of the backing memory is managed
+/// elsewhere. Currently used with the pooling allocator.
+pub(crate) struct ExternalMemory {
+    /// The memory in the host for this wasm memory. The length of this
+    /// slice is the maximum size of the memory that can be grown to.
+    base: &'static mut [u8],
 
-        /// The current size, in bytes, of this memory.
-        size: usize,
+    /// The current size, in bytes, of this memory.
+    size: usize,
 
-        /// A callback which makes portions of `base` accessible for when memory
-        /// is grown. Otherwise it's expected that accesses to `base` will
-        /// fault.
-        make_accessible: Option<fn(*mut u8, usize) -> Result<()>>,
+    /// A callback which makes portions of `base` accessible for when memory
+    /// is grown. Otherwise it's expected that accesses to `base` will
+    /// fault.
+    make_accessible: Option<fn(*mut u8, usize) -> Result<()>>,
+
+    /// The image management, if any, for this memory. Owned here and
+    /// returned to the pooling allocator when termination occurs.
+    pub(crate) memory_image: Option<MemoryImageSlot>,
+}
 
-        /// The image management, if any, for this memory. Owned here and
-        /// returned to the pooling allocator when termination occurs.
+impl ExternalMemory {
+    fn new(
+        base: &'static mut [u8],
+        size: usize,
+        make_accessible: Option<fn(*mut u8, usize) -> Result<()>>,
         memory_image: Option<MemoryImageSlot>,
-    },
+    ) -> Result<Self> {
+        if base.len() < size {
+            bail!(
+                "initial memory size of {} exceeds the pooling allocator's \
+                 configured maximum memory size of {} bytes",
+                size,
+                base.len(),
+            );
+        }
 
-    /// A "dynamic" memory whose data is managed at runtime and lifetime is tied
-    /// to this instance.
-    Dynamic(Box<dyn RuntimeLinearMemory>),
+        if let Some(make_accessible) = make_accessible {
+            if size > 0 {
+                make_accessible(base.as_mut_ptr(), size)?;
+            }
+        }
+
+        Ok(Self {
+            base,
+            size,
+            make_accessible,
+            memory_image,
+        })
+    }
+}
+
+impl Default for ExternalMemory {
+    fn default() -> Self {
+        Self {
+            base: &mut [],
+            size: 0,
+            make_accessible: Some(|_, _| unreachable!()),
+            memory_image: None,
+        }
+    }
 }
 
+impl RuntimeLinearMemory for ExternalMemory {
+    fn byte_size(&self) -> usize {
+        self.size
+    }
+
+    fn maximum_byte_size(&self) -> Option<usize> {
+        Some(self.base.len())
+    }
+
+    fn grow_to(&mut self, new_byte_size: usize) -> Result<()> {
+        // Never exceed the static memory size.
+        if new_byte_size > self.base.len() {
```

Comment:
I think this can be an assert since this should have already been checked in `Memory::grow`

Reply:
(I think this was also present in the original code where this check wasn't necessary)

Reply:
I see other checks in `Memory::grow` but none that are equivalent to this one. Do you mean that we should move this check back to the `Memory::grow` function and duplicate it here and in `MmapMemory` as a `debug_assert!`?

Reply:
I think it's covered by [this check](https://github.com/bytecodealliance/wasmtime/blob/128c42fa0931611b620021fee69a5f2c2c2c0bb1/crates/runtime/src/memory.rs#L519-L525) since this variant always has a `maximum_byte_size` of `Some`, but yeah here I think this check can be an `assert!` or a `debug_assert!`